silk-react-components/src/HierarchicalMapping/Components/MappingsWorkview.jsx

```diff
@@ -25,29 +20,83 @@ import {
     MAPPING_RULE_TYPE_OBJECT,
     MAPPING_RULE_TYPE_ROOT,
 } from '../helpers';
+import { MESSAGES } from '../constants';
+import EventEmitter from '../utils/EventEmitter';
 
-const MappingsWorkview = React.createClass({
-    mixins: [UseMessageBus],
-
+class MappingsWorkview extends React.Component {
+    
     // define property types
-    propTypes: {
-        currentRuleId: React.PropTypes.string, // selected rule id
-    },
-    onRuleCreate({type}) {
+    static propTypes = {
+        onToggleTreeNav: PropTypes.func,
+        onRuleIdChange: PropTypes.func,
+        onAskDiscardChanges: PropTypes.func,
+        onClickedRemove: PropTypes.func,
+        currentRuleId: PropTypes.string, // selected rule id
+        askForDiscardData: PropTypes.oneOfType([PropTypes.object, PropTypes.bool]), // selected rule id
+    };
+    
+    state = {
+        loading: true,
+        ruleData: {},
+        ruleEditView: false,
+        editing: [],
+        isCopying: !!sessionStorage.getItem('copyingData'),
+        showSuggestions: false,
+        askForChilds: false,
+    };
+    
+    componentDidMount() {
+        this.loadData({ initialLoad: true });
+        EventEmitter.on(MESSAGES.RELOAD, this.loadData);
+        EventEmitter.on(MESSAGES.RULE_ID.CREATE, this.onRuleCreate);
+        EventEmitter.on(MESSAGES.MAPPING.CREATE, this.handleCreate);
+        EventEmitter.on(MESSAGES.MAPPING.SHOW_SUGGESTIONS, this.handleShowSuggestions);
+        EventEmitter.on(MESSAGES.RULE_VIEW.UNCHANGED, this.handleRuleEditClose);
+        EventEmitter.on(MESSAGES.RULE_VIEW.CLOSE, this.handleRuleEditClose);
+        EventEmitter.on(MESSAGES.RULE_VIEW.CHANGE, this.handleRuleEditOpen);
+        EventEmitter.on(MESSAGES.RULE_VIEW.DISCARD_ALL, this.discardAll);
+    }
+    
+    componentWillUnmount() {
+        EventEmitter.off(MESSAGES.RELOAD, this.loadData);
+        EventEmitter.off(MESSAGES.RULE_ID.CREATE, this.onRuleCreate);
+        EventEmitter.off(MESSAGES.MAPPING.CREATE, this.handleCreate);
+        EventEmitter.off(MESSAGES.MAPPING.SHOW_SUGGESTIONS, this.handleShowSuggestions);
+        EventEmitter.off(MESSAGES.RULE_VIEW.UNCHANGED, this.handleRuleEditClose);
+        EventEmitter.off(MESSAGES.RULE_VIEW.CLOSE, this.handleRuleEditClose);
+        EventEmitter.off(MESSAGES.RULE_VIEW.CHANGE, this.handleRuleEditOpen);
+        EventEmitter.off(MESSAGES.RULE_VIEW.DISCARD_ALL, this.discardAll);
+    }
+    
+    componentDidUpdate(prevProps, prevState) {
+        if (prevProps.currentRuleId !== this.props.currentRuleId) {
+            this.loadData();
+        }
+    }
+    
+    shouldComponentUpdate(nextProps, nextState) {
+        // Required to prevent empty redraws while not all data is there.
+        // The issue is due to bad use of React ...
```

Comment:
still bad use or can it be fixed?

Reply:
I removed `shouldComponentUpdate` and all working fine, without any issues, I can't figure out why we need it, maybe it fixed some kind of issue which already fixed

Reply:
👍